### PR TITLE
[SGD] Better support for custom DDP

### LINKED ISSUE #11705: [SGD] Better support for custom DDP
Right now it's difficult to customize the DDP wrapping for the TorchTrainer. We should 
1) Allow users to pass in custom args to DDP (create a `ddp_args` similar to `apex_args`).
2) Provide better documentation on how to use custom Distributed training wrappers.

### Describe your feature request

For feature requests or questions, post on our Github Discussions page instead: https://github.com/ray-project/ray/discussions!

### PR DESCRIPTION
**Pull request by amogkam**
## Why are these changes needed?
* Add ddp_args field to register and move apex_args to register as well
* Update documentation

## Related issue number

Closes https://github.com/ray-project/ray/issues/11705, https://github.com/ray-project/ray/issues/10096

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by richardliaw** [comment]
@krfricke could you take a look? thanks!

File: python/ray/util/sgd/torch/torch_runner.py [hunk 1]
```diff
@@ -40,7 +39,6 @@ def __init__(self,
         self.use_gpu = use_gpu
         self.use_fp16 = use_fp16
         self.use_tqdm = use_tqdm
-        self.apex_args = apex_args or {}
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_runner.py:43]: Can you leave a deprecation message if the user passes this in?

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 2]
```diff
@@ -150,7 +146,6 @@ def __init__(
             timeout_s=NCCL_TIMEOUT_S,
             use_fp16=False,
             use_tqdm=False,
-            apex_args=None,
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_trainer.py:153]: Can you leave a deprecation message if the user passes this in?

File: python/ray/util/sgd/torch/torch_trainer.py [hunk 3]
```diff
@@ -150,7 +146,6 @@ def __init__(
             timeout_s=NCCL_TIMEOUT_S,
             use_fp16=False,
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/torch_trainer.py:147]: Should we move `use_fp16` also into `register`?

**Review by richardliaw — APPROVED** [review]
looks great! left a couple nits

File: python/ray/util/sgd/torch/torch_trainer.py (hunk 3 above)
Comment (amogkam) [reply python/ray/util/sgd/torch/torch_trainer.py:147]: I think it's fine to leave it as is. Currently the TorchRunner needs to be aware of this to know whether to save/load state from amp, so it needs to be passed in via the TorchTrainer.

File: python/ray/util/sgd/torch/torch_trainer.py (hunk 2 above)
Comment (amogkam) [reply python/ray/util/sgd/torch/torch_trainer.py:153]: Yes good point; done

**Comment by richardliaw** [comment]
Hmm, sgd test is failing

**Comment by amogkam** [comment]
@richardliaw fixed now- tests are passing.

**Pull request merged**